File: PathoFroh/src/main/java/com/patho/main/model/Council.java
package com.patho.main.model;

import java.util.Date;
import java.util.Set;

import javax.persistence.Column;
import javax.persistence.Embedded;
import javax.persistence.Entity;
import javax.persistence.EntityListeners;
import javax.persistence.EnumType;
import javax.persistence.Enumerated;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.OneToOne;
import javax.persistence.OrderBy;
import javax.persistence.SequenceGenerator;
import javax.persistence.Temporal;
import javax.persistence.TemporalType;
import javax.persistence.Transient;
import javax.persistence.Version;

import org.hibernate.annotations.DynamicUpdate;
import org.hibernate.annotations.SelectBeforeUpdate;
import org.hibernate.envers.Audited;

import com.patho.main.common.CouncilState;
import com.patho.main.model.interfaces.DataList;
import com.patho.main.model.interfaces.ID;
import com.patho.main.model.patient.Task;
import com.patho.main.model.util.audit.Audit;
import com.patho.main.model.util.audit.AuditAble;
import com.patho.main.model.util.audit.AuditListener;
import com.patho.main.util.helper.TextToLatexConverter;

import lombok.Getter;
import lombok.Setter;

@Entity
@Audited
@SelectBeforeUpdate(true)
@DynamicUpdate(true)
@SequenceGenerator(name = "council_sequencegenerator", sequenceName = "council_sequence")
@Getter
@Setter
@EntityListeners(AuditListener.class)
public class Council implements ID, DataList, AuditAble {

	@Id
	@GeneratedValue(generator = "council_sequencegenerator")
	@Column(unique = true, nullable = false)
	private long id;

	@Version
	private long version;

	@Embedded
	protected Audit audit;

	@ManyToOne(fetch = FetchType.LAZY)
	private Task task;

	/**
	 * Name of the council
	 */
	@Column(columnDefinition = "VARCHAR")
	private String name;

	/**
	 * Council physician
	 */
	@OneToOne
	private Physician councilPhysician;

	/**
	 * Physician to sign the council
	 */
	@OneToOne
	private Physician physicianRequestingCouncil;

	/**
	 * Date of request
	 */
	@Temporal(TemporalType.DATE)
	private Date dateOfRequest;

	/**
	 * Text of council
	 */
	@Column(columnDefinition = "text")
	private String councilText;

	/**
	 * True if samples were send to external clinics
	 */
	@Column
	private boolean councilRequestCompleted;

	/**
	 * State of the council
	 */
	@Enumerated(EnumType.STRING)
	private CouncilNotificationMethod notificationMethod;

	/**
	 * True if samples were send to external clinics
	 */
	@Column
	private boolean sampleShipped;

	/**
	 * Attached slides of the council
	 */
	@Column(columnDefinition = "text")
	private String sampleShippedCommentary;

	/**
	 * Date of request
	 */
	@Temporal(TemporalType.DATE)
	private Date sampleShippedDate;

	/**
	 * True if the samples should be returned
	 */
	@Column
	private boolean expectSampleReturn;

	/**
	 * True if sample is returned
	 */
	@Column
	private boolean sampleReturned;

	/**
	 * True if sample is returned
	 */
	@Column
	private boolean councilCompleted;

	/**
	 * Date of request
	 */
	@Temporal(TemporalType.DATE)
	private Date councilCompletedDate;

	/**
	 * Commentary
	 */
	@Column(columnDefinition = "text")
	private String sampleReturnedCommentary;

	/**
	 * Date of request
	 */
	@Temporal(TemporalType.DATE)
	private Date sampleReturnedDate;

	/**
	 * Commentary
	 */
	@Column(columnDefinition = "text")
	private String commentary;

	/**
	 * True if answer was received
	 */
	@Column
	private boolean replyReceived;

	/**
	 * Date of return
	 */
	@Temporal(TemporalType.DATE)
	private Date replyReceivedDate;

	/**
	 * Pdf attached to this council
	 */
	@OneToMany(fetch = FetchType.LAZY)
	@OrderBy("audit.createdOn DESC")
	private Set<PDFContainer> attachedPdfs;

	public Council() {
	}

	public Council(Task task) {
		this.task = task;
	}

	@Transient
	public String getCouncilTextAsLatex() {
		return (new TextToLatexConverter()).convertToTex(getCouncilText());
	}

	@Override
	@Transient
	public String getPublicName() {
		return "Konsil - " + task.getTaskID();
	}

	@Override
	public boolean equals(Object obj) {
		if (obj instanceof Council && ((Council) obj).getId() == getId())
			return true;
		return super.equals(obj);
	}

	/**
	 * Method of notification
	 * 
	 * @author andi
	 *
	 */
	public static enum CouncilNotificationMethod {
		MTA, SECRETARY, NONE;
	}
}

File: src/main/kotlin/com/patho/main/model/patient/miscellaneous/Council.kt
package com.patho.main.model.patient.miscellaneous

import com.patho.main.model.AbstractPersistable
import com.patho.main.model.PDFContainer
import com.patho.main.model.Physician
import com.patho.main.model.audit.AuditAble
import com.patho.main.model.interfaces.DataList
import com.patho.main.model.patient.Patient
import com.patho.main.model.patient.Task
import com.patho.main.model.util.audit.Audit
import com.patho.main.model.util.audit.AuditListener
import com.patho.main.util.helper.TextToLatexConverter
import org.hibernate.annotations.SelectBeforeUpdate
import org.hibernate.envers.Audited
import java.time.LocalDate
import java.util.*
import javax.persistence.*
import kotlin.collections.HashSet

@Entity
@Audited
@SelectBeforeUpdate(true)
@SequenceGenerator(name = "council_sequencegenerator", sequenceName = "council_sequence")
@EntityListeners(AuditListener::class)
open class Council : AbstractPersistable, DataList, AuditAble {

    @Id
    @GeneratedValue(generator = "council_sequencegenerator")
    @Column(unique = true, nullable = false)
    open override var id: Long = 0

    @Version
    open var version: Long = 0

    @Embedded
    open override var audit: Audit? = null

    @ManyToOne(fetch = FetchType.LAZY)
    open var task: Task? = null

    /**
     * Name of the council
     */
    @Column(columnDefinition = "VARCHAR")
    open var name: String = ""

    /**
     * Council physician
     */
    @OneToOne
    open var councilPhysician: Physician? = null

    /**
     * Physician to sign the council
     */
    @OneToOne
    open var physicianRequestingCouncil: Physician? = null

    /**
     * Date of request
     */
    open var dateOfRequest: LocalDate? = null

    /**
     * Text of council
     */
    @Column(columnDefinition = "text")
    open var councilText: String? = null

    /**
     * True if samples were send to external clinics
     */
    @Column
    open var councilRequestCompleted: Boolean = false

    /**
     * Date of request completed
     */
    open var councilRequestCompletedDate: LocalDate? = null

    /**
     * State of the council
     */
    @Enumerated(EnumType.STRING)
    open var notificationMethod: CouncilNotificationMethod? = null

    /**
     * True if samples were send to external clinics
     */
    @Column
    open var sampleShipped: Boolean = false

    /**
     * Date of request
     */
    open var sampleShippedDate: LocalDate? = dateOfRequest

    /**
     * Attached slides of the council
     */
    @Column(columnDefinition = "text")
    open var sampleShippedCommentary: String? = null

    /**
     * True if the samples should be returned
     */
    @Column
    open var expectSampleReturn: Boolean = false

    /**
     * True if sample is returned
     */
    @Column
    open var sampleReturned: Boolean = false

    /**
     * Date of request
     */
    open var sampleReturnedDate: LocalDate? = null

    /**
     * Commentary
     */
    @Column(columnDefinition = "text")
    open var sampleReturnedCommentary: String? = null

    /**
     * True if answer was received
     */
    @Column
    open var replyReceived: Boolean = false

    /**
     * Date of return
     */
    open var replyReceivedDate: LocalDate? = null

    /**
     * True if sample is returned
     */
    @Column
    open var councilCompleted: Boolean = false

    /**
     * Date of request
     */
    open var councilCompletedDate: LocalDate? = null

    /**
     * Commentary
     */
    @Column(columnDefinition = "text")
    open var commentary: String? = null

    /**
     * Pdf attached to this council
     */
    @OneToMany(fetch = FetchType.LAZY)
    @OrderBy("audit.createdOn DESC")
    open override var attachedPdfs: MutableSet<PDFContainer> = HashSet<PDFContainer>()

    constructor()

    constructor(task: Task) {
        this.task = task
    }

    @Transient
    open fun getCouncilTextAsLatex(): String {
        return TextToLatexConverter().convertToTex(councilText)
    }

    @Transient
    open fun isShippentExpected(): Boolean {
        return notificationMethod != CouncilNotificationMethod.NONE
    }

    override val publicName: String
        @Transient
        get() = "Konsil - " + task?.taskID

    override val patient: Patient?
        @Transient
        get() = task?.patient

    /**
     * Method of notification
     *
     * @author andi
     */
    enum class CouncilNotificationMethod {
        MTA, SECRETARY, NONE
    }
}
